include/dxc/Support/dxcapi.use.h

```diff
@@ -13,38 +13,27 @@
 #define __DXCAPI_USE_H__
 
 #include "dxc/dxcapi.h"
+#include "llvm/Support/DynamicLibrary.h"
 
 namespace dxc {
 
 // Helper class to dynamically load the dxcompiler or a compatible libraries.
 class DxcDllSupport {
 protected:
-  HMODULE m_dll;
+  using DynamicLibrary = llvm::sys::DynamicLibrary;
+  DynamicLibrary m_dll;
   DxcCreateInstanceProc m_createFn;
   DxcCreateInstance2Proc m_createFn2;
 
-  #ifndef _WIN32
-  void FreeLibrary(void* handle) {
-    ::dlclose(handle);
-  }
-  HMODULE LoadLibraryW(LPCWSTR name) {
-    return ::dlopen(CW2A(name).m_psz, RTLD_LAZY);
-  }
-  HMODULE GetProcAddress(HMODULE dll, LPCSTR fnName) {
-    return ::dlsym(dll, fnName);
-  }
-  #endif
-
   HRESULT InitializeInternal(LPCWSTR dllName, LPCSTR fnName) {
-    if (m_dll != nullptr) return S_OK;
-    m_dll = LoadLibraryW(dllName);
-    if (m_dll == nullptr) return HRESULT_FROM_WIN32(GetLastError());
-    m_createFn = (DxcCreateInstanceProc)GetProcAddress(m_dll, fnName);
+    if (m_dll.isValid()) return S_OK;
+    m_dll = DynamicLibrary::getPermanentLibrary(CW2A(dllName));
+    if (!m_dll.isValid()) return E_FAIL;
```

Comment:
This was the cause of the automated testing failure. Getpermanentlibrary makes calls that reset the windows error so this returns S_OK even when the library fails to load.

Reply:
This was the cause of the automated testing failure. Getpermanentlibrary makes calls that reset the windows error so this returns S_OK even when the library fails to load.

Reply:
In that case, I think instead of returning E_FAIL, maybe we should return something more meaningful; closer to the original implementation.
e.g.
```cpp
return HRESULT_FROM_WIN32(ERROR_DLL_INIT_FAILED);
```
or
```cpp
return HRESULT_FROM_WIN32(ERROR_DLL_NOT_FOUND);
```
or
...
This reference has error codes related to DLLs: 
https://msdn.microsoft.com/en-us/library/windows/desktop/ms681383(v=vs.85).aspx

I suppose ERROR_DLL_INIT_FAILED should be okay. (we're not able to check for dll existence here i suppose).

Reply:
What do you think?